File: apps/processor/src/embedding.ts
import type { EmbeddingResult } from "./types.js";

const LOCAL_EMBEDDING_MODEL = "local-embeddings";
const LOCAL_EMBEDDING_DIMENSIONS = 384;

interface EmbeddingApiResponse {
	data?: Array<{ embedding: number[] }>;
	model?: string;
}

export async function fetchEmbedding(
	endpoint: string,
	model: string,
	input: string,
): Promise<EmbeddingResult> {
	if (!endpoint && model === LOCAL_EMBEDDING_MODEL) {
		return {
			model,
			vector: embedText(input),
		};
	}
	if (!endpoint) {
		throw new Error("EMBEDDING_ENDPOINT is required for non-local embedding models");
	}

	const response = await fetch(endpoint, {
		method: "POST",
		headers: {
			"Content-Type": "application/json",
		},
		body: JSON.stringify({
			model,
			input,
		}),
	});

	if (!response.ok) {
		throw new Error(`embedding request failed: ${response.status} ${response.statusText}`);
	}

	const payload = (await response.json()) as EmbeddingApiResponse;
	const vector = payload.data?.[0]?.embedding;
	if (!vector) {
		throw new Error("embedding response did not include a vector");
	}

	return {
		model: payload.model ?? model,
		vector,
	};
}

export function embedText(input: string): number[] {
	const vector = Array.from({ length: LOCAL_EMBEDDING_DIMENSIONS }, () => 0);
	const tokens = tokenize(input);
	if (tokens.length === 0) {
		tokens.push(input.trim().toLowerCase());
	}

	for (const token of tokens) {
		const hash = fnv1a(token);
		const index = hash % LOCAL_EMBEDDING_DIMENSIONS;
		vector[index] += (hash & 0x80000000) !== 0 ? -1 : 1;
	}

	normalize(vector);
	return vector;
}

function tokenize(input: string): string[] {
	return input.toLowerCase().match(/[a-z0-9]+/g) ?? [];
}

function fnv1a(value: string): number {
	let hash = 2166136261;
	for (let index = 0; index < value.length; index += 1) {
		hash ^= value.charCodeAt(index);
		hash = Math.imul(hash, 16777619) >>> 0;
	}
	return hash;
}

function normalize(vector: number[]): void {
	const magnitude = Math.sqrt(vector.reduce((sum, value) => sum + value * value, 0));
	if (magnitude === 0) {
		return;
	}

	for (let index = 0; index < vector.length; index += 1) {
		vector[index] /= magnitude;
	}
}

File: apps/pkg/embeddingutil/embedding.go
package embeddingutil

import (
	"math"
	"strings"
)

const (
	DefaultModel = "local-embeddings"
	Dimensions   = 384
)

func EmbedText(input string) []float64 {
	vector := make([]float64, Dimensions)
	tokens := tokenize(input)
	if len(tokens) == 0 {
		tokens = []string{strings.ToLower(strings.TrimSpace(input))}
	}

	for _, token := range tokens {
		hash := fnv1a(token)
		index := int(hash % Dimensions)
		weight := 1.0
		if hash&0x80000000 != 0 {
			weight = -1.0
		}
		vector[index] += weight
	}

	normalize(vector)
	return vector
}

func tokenize(input string) []string {
	return strings.FieldsFunc(strings.ToLower(input), func(r rune) bool {
		return !((r >= 'a' && r <= 'z') || (r >= '0' && r <= '9'))
	})
}

func fnv1a(value string) uint32 {
	var hash uint32 = 2166136261
	for i := 0; i < len(value); i++ {
		hash ^= uint32(value[i])
		hash *= 16777619
	}
	return hash
}

func normalize(vector []float64) {
	var sum float64
	for _, value := range vector {
		sum += value * value
	}
	if sum == 0 {
		return
	}

	magnitude := math.Sqrt(sum)
	for index := range vector {
		vector[index] /= magnitude
	}
}
